source/extensions/filters/http/sxg/filter.cc

```diff
@@ -0,0 +1,499 @@
+#include "source/extensions/filters/http/sxg/filter.h"
+
+#include <cstdlib>
+#include <regex>
+#include <string>
+#include <chrono>
+
+#include <openssl/evp.h>
+#include <openssl/bio.h>
+#include <openssl/pem.h>
+#include <openssl/x509.h>
+
+#include "envoy/server/filter_config.h"
+
+#include "absl/strings/escaping.h"
+#include "envoy/stats/scope.h"
+#include "envoy/http/codes.h"
+#include "source/common/common/utility.h"
+#include "source/common/http/headers.h"
+#include "source/common/stats/utility.h"
+
+#include "libsxg.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace SXG {
+
+Http::RegisterCustomInlineHeader<Http::CustomInlineHeaderRegistry::Type::RequestHeaders>
+    accept_handle(Http::CustomHeaders::get().Accept);
+
+template <class T>
+const std::vector<std::string> initializeHeaderPrefixFilters(const T& filters_proto) {
+  std::vector<std::string> filters;
+  filters.reserve(filters_proto.size());
+
+  for (const auto& filter : filters_proto) {
+    filters.emplace_back(filter);
+  }
+
+  return filters;
+}
+
+FilterConfig::FilterConfig(const envoy::extensions::filters::http::sxg::v3alpha::SXG& proto_config,
+                           TimeSource& time_source, std::shared_ptr<SecretReader> secret_reader,
+                           const std::string& stat_prefix, Stats::Scope& scope)
+    : stats_(generateStats(stat_prefix + "sxg.", scope)),
+      duration_(proto_config.has_duration() ? proto_config.duration().seconds() : 604800UL),
+      cbor_url_(proto_config.cbor_url()), validity_url_(proto_config.validity_url()),
+      mi_record_size_(proto_config.mi_record_size() ? proto_config.mi_record_size() : 4096L),
+      client_can_accept_sxg_header_((proto_config.client_can_accept_sxg_header().length() > 0)
+                                        ? proto_config.client_can_accept_sxg_header()
+                                        : "x-envoy-client-can-accept-sxg"),
+      should_encode_sxg_header_((proto_config.should_encode_sxg_header().length() > 0)
+                                    ? proto_config.should_encode_sxg_header()
+                                    : "x-envoy-should-encode-sxg"),
+      header_prefix_filters_(initializeHeaderPrefixFilters(proto_config.header_prefix_filters())),
+      time_source_(time_source), secret_reader_(secret_reader) {}
+
+const std::regex& matchAcceptHeaderRegex() {
+  // RegEx to detect Signed Exchange Accept Headers from:
+  // https://web.dev/signed-exchanges/#best-practices
+  static const auto regex = new std::regex("(?:^|,)\\s*(application/signed-exchange|text/"
+                                           "html)\\s*(?:;v=([\\w]+))?\\s*(?:;q=([\\d\\.]+))?\\s*",
+                                           std::regex_constants::icase);
```

Comment:
I think we want `CONSTRUCT_ON_FIRST_USE()` for this regex to avoid the static init